modules/prebid/ortb2blocking/hook_raw_bidder_response.go

```diff
@@ -0,0 +1,384 @@
+		if err != nil {
+			return failedChecksData, fmt.Errorf("failed to get override for battr.allowed_banner_attr_for_deals: %s", err)
+		}
+	}
+
+	if blockAttr.bAttr == nil || len(blockAttr.bAttr) == 0 {
+		return failedChecksData, nil
+	}
+	blockedBattr := blockAttr.bAttr[bid.ImpID]
+	bidAttr := toInt(bid.Attr)
+
+	shouldBlock, blockedAttributes := blockAttribute(bidAttr, blockedBattr, dealExceptions, false)
+	if shouldBlock {
+		failedChecksData["battr"] = blockedAttributes
+	}
+
+	return failedChecksData, nil
+}
+
+func blockAttribute[Attribute comparable](attributes, blockedAttributes, dealExceptions []Attribute, blockUnknown bool) (bool, []Attribute) {
+	if len(attributes) == 0 {
+		if blockUnknown {
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if len(blockedAttributes) == 0 {
+		return false, nil
+	}
+
+	blockedElements := make([]Attribute, 0)
+	shouldBlock := true
+	for _, attribute := range attributes {
+		for _, blockedAttribute := range blockedAttributes {
+			if attribute == blockedAttribute {
+				for _, allowedAttribute := range dealExceptions {
+					if attribute == allowedAttribute {
+						shouldBlock = false
+						break
+					}
+				}
+				if shouldBlock {
+					blockedElements = append(blockedElements, attribute)
+				}
+			}
+			shouldBlock = true
+		}
+	}
+
+	if len(blockedElements) == 0 {
+		return false, nil
+	}
+
+	return true, blockedElements
+}
+
+func findDealExceptionsOverridesOrDefault[T any](
+	dealID string,
+	overrideGetter overrideGetterFn[[]T],
+	actionOverrides []ActionOverride,
+	defaultOverride []T,
+) (overrides []T, err error) {
+	for _, action := range actionOverrides {
+		if action.Conditions.DealIds == nil {
+			return overrides, errors.New("conditions field in account configuration must contain deal_ids")
+		}
+
+		for _, id := range action.Conditions.DealIds {
+			if id == dealID {
+				actionOverride, err := overrideGetter(action.Override)
+				if err != nil {
+					return overrides, err
+				}
+				overrides = append(overrides, actionOverride...)
+			}
+		}
+	}
+
+	if len(overrides) == 0 {
+		return defaultOverride, nil
+	}
+
+	overrides = append(overrides, defaultOverride...)
+
+	return overrides, nil
+}
+
+func addDebugMessage(
+	result *hookstage.HookResult[hookstage.RawBidderResponsePayload],
+	bid *openrtb2.Bid,
+	bidder string,
+	failedAttributes string,
+) {
+	result.DebugMessages = append(
+		result.DebugMessages,
+		fmt.Sprintf("Bid %s from bidder %s has been rejected, failed checks: %s", bid.ID, bidder, failedAttributes),
+	)
+}
+
+// returns a slice with the names of failed attributes
+func getFailedAttributes(data map[string]interface{}) string {
+	var builder strings.Builder
+	for i, attribute := range [5]string{
+		"badv",
+		"bcat",
+		"cattax",
+		"bapp",
+		"battr",
+	} {
+		if _, ok := data[attribute]; ok {
+			if i > 0 && builder.Len() != 0 {
+				builder.WriteString(", ")
+			}
+			builder.WriteString(attribute)
+		}
+	}
+
+	return builder.String()
+}
```

Comment:
I'm not sure I see the reason for returning a comma-delimited string here. I see that this function is used for analytics and debugging. It appears that the failed attributes should be made available to analytics modules as an array of strings while the failed attributes should be listed as a comma-separated string in debug info.
In that case, can we make this function return a slice of attributes instead of a comma-delimited string? Then the slice can be added to analytics without further modification and we just run `strings.Join` to concatenate the results for debug in `addDebugMessage`.
```
// returns a slice with the names of failed attributes
func getFailedAttributes(data map[string]interface{}) []string {
	var builder []string
	for _, attribute := range [5]string{
		"badv",
		"bcat",
		"cattax",
		"bapp",
		"battr",
	} {
		if _, ok := data[attribute]; ok {
			builder = append(builder, attribute)
		}
	}

	return builder
}
```

Thoughts?

Reply:
I agree, this is simpler, thanks for the working example. I overlooked that adding to the slice by index in the original implementation is not necessary, as we are already iterating over a sorted slice of attributes and we can append without worries. Updated.